File: S3DE/Graphics/Meshes/StandardMesh.cs
using S3DE.Maths;
using S3DE.Utility;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S3DE.Graphics.Meshes
{
    public class StandardMesh : Mesh
    {
        public Vector3[] Vertices { private get; set; }
        public Vector2[] Uvs { private get; set; }
        public ushort[] Indicies { private get; set; }

        public StandardMesh() : base()
        {
            SetVertexAttribute(new VertexAttribute(0, 3, GLType.FLOAT, false, 20, 0));
            SetVertexAttribute(new VertexAttribute(1, 2, GLType.FLOAT, false, 20, 12));
            EnableVertexAttribute(0);
            EnableVertexAttribute(1);
        }

        public static StandardMesh CreateCube(Vector3 scale)
        {
            StandardMesh m = new StandardMesh();

            float vX = 0.5f * scale.x;
            float vY = 0.5f * scale.y;
            float vZ = 0.5f * scale.z;

            m.Vertices = new Vector3[] {new Vector3(-0.5f, -0.5f, -0.5f), new Vector3(-0.5f, 0.5f, -0.5f), new Vector3(0.5f, 0.5f, -0.5f),new Vector3(0.5f,-0.5f,-0.5f),
                                        new Vector3(-0.5f,0.5f,-0.5f),new Vector3(-0.5f,0.5f,0.5f),new Vector3(0.5f,0.5f,0.5f),new Vector3(0.5f,0.5f,-0.5f),
                                        new Vector3(0.5f,-0.5f,-0.5f),new Vector3(0.5f,0.5f,-0.5f),new Vector3(0.5f,0.5f,0.5f),new Vector3(0.5f,-0.5f,0.5f),
                                        new Vector3(-0.5f,-0.5f,0.5f),new Vector3(-0.5f,0.5f,0.5f),new Vector3(-0.5f,0.5f,-0.5f),new Vector3(-0.5f,-0.5f,-0.5f),
                                        new Vector3(0.5f,-0.5f,0.5f),new Vector3(0.5f,0.5f,0.5f),new Vector3(-0.5f,0.5f,0.5f),new Vector3(-0.5f,-0.5f,0.5f),
                                        new Vector3(-0.5f,-0.5f,0.5f),new Vector3(-0.5f,-0.5f,-0.5f),new Vector3(0.5f,-0.5f,-0.5f),new Vector3(0.5f,-0.5f,0.5f)};

            m.Uvs = new Vector2[] { new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0),
                                    new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0),
                                    new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0),
                                    new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0),
                                    new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0),
                                    new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0)};


            m.Indicies = new ushort[] { 0,1,2,2,3,0,
                                        4,5,6,6,7,4,
                                        8,9,10,10,11,8,
                                        12,13,14,14,15,12,
                                        16,17,18,18,19,16,
                                        20,21,22,22,23,20};
            m.Apply();

            return m;
        }

        public override void Apply()
        {
            using (ByteBuffer V_BB = ByteBuffer.Create(Vertices.Length * 4))
            {

                for (int i = 0; i < Vertices.Length; i++)
                {
                    Vector3 v = Vertices[i];
                    Vector2 uv = Uvs[i];
                    V_BB.AddRange(4, v.x, v.y, v.z);
                    V_BB.AddRange(4, uv.x, uv.y);
                }

                UploadMeshData(V_BB.Data, Indicies);
            }
        }
    }
}

File: S3DECore/StandardMesh.cpp
#include "Mesh.h"
#include "Renderer.h"

namespace S3DECore {
	namespace Graphics {
		namespace Meshes {

			StandardMesh::StandardMesh() : Mesh() {
				Bind();
				SetVertexAttrib(0, 3, GL_FLOAT, GL_FALSE, 20, 0);
				SetVertexAttrib(1, 2, GL_FLOAT, GL_FALSE, 20, 12);
				EnableVertexAttrib(0);
				EnableVertexAttrib(1);
			}

			StandardMesh::~StandardMesh() {
				
			}

			StandardMesh::!StandardMesh() {

			}

			void StandardMesh::Apply() {
				array<float>^ vertData = gcnew array<float>((Vertices->Length * 3) + (uvs->Length * 2));
				
				//Copy the vertices to a byte array.
				for (int i = 0; i < verts->Length; i++) {
					int j = i * 5;
					vertData[j] = Vertices[i].x;
					vertData[j + 1] = Vertices[i].y;
					vertData[j + 2] = Vertices[i].z;
					vertData[j + 3] = uvs[i].x;
					vertData[j + 4] = uvs[i].y;
				}
				
				pin_ptr<float> vertPtr(&vertData[0]);
				SetVertexData(vertPtr, vertData->Length, GL_STATIC_DRAW);

				pin_ptr<ushort> indPtr(&indicies[0]);
				SetIndicies(indPtr, indicies->Length, GL_STATIC_DRAW);
			}

			StandardMesh^ StandardMesh::CreateCube(Vector3 scale) {
				StandardMesh^ m = gcnew StandardMesh();

				float vX = 0.5f * scale.x;
				float vY = 0.5f * scale.y;
				float vZ = 0.5f * scale.z;

				array<Vector3>^ verts = {   Vector3(-0.5f, -0.5f, -0.5f), Vector3(-0.5f, 0.5f, -0.5f), Vector3(0.5f, 0.5f, -0.5f),Vector3(0.5f,-0.5f,-0.5f),
											Vector3(-0.5f,0.5f,-0.5f),	  Vector3(-0.5f,0.5f,0.5f),	   Vector3(0.5f,0.5f,0.5f),   Vector3(0.5f,0.5f,-0.5f),
											Vector3(0.5f,-0.5f,-0.5f),	  Vector3(0.5f,0.5f,-0.5f),	   Vector3(0.5f,0.5f,0.5f),   Vector3(0.5f,-0.5f,0.5f),
											Vector3(-0.5f,-0.5f,0.5f),	  Vector3(-0.5f,0.5f,0.5f),	   Vector3(-0.5f,0.5f,-0.5f), Vector3(-0.5f,-0.5f,-0.5f),
											Vector3(0.5f,-0.5f,0.5f),	  Vector3(0.5f,0.5f,0.5f),	   Vector3(-0.5f,0.5f,0.5f),  Vector3(-0.5f,-0.5f,0.5f),
											Vector3(-0.5f,-0.5f,0.5f),	  Vector3(-0.5f,-0.5f,-0.5f),  Vector3(0.5f,-0.5f,-0.5f), Vector3(0.5f,-0.5f,0.5f)};

				m->Vertices = verts;

				array<Vector2>^ uvs = { Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0),
										Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0),
										Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0),
										Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0),
										Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0),
										Vector2(0, 0), Vector2(0, 1), Vector2(1, 1), Vector2(1, 0) };

				m->Uvs = uvs;

				array<ushort>^ inds =	  { 0,1,2,2,3,0,
											4,5,6,6,7,4,
											8,9,10,10,11,8,
											12,13,14,14,15,12,
											16,17,18,18,19,16,
											20,21,22,22,23,20 };


				m->Indicies = inds;

				m->Apply();

				return m;
			}
		}
	}
}